### ISSUE TITLE
Mutating objects affect reproducibility

### ISSUE THREAD

**Issue by benjreinhart** [marimo-team/marimo#1383]
### Describe the bug

In [FAQ](https://docs.marimo.io/faq.html#what-problems-does-marimo-solve), Marimo states:

> In contrast, marimo guarantees that your code, outputs, and program state are consistent, eliminating hidden state and making your notebook reproducible.

I see that variables are tracked and reassignment is not allowed. However, since Python is not an immutable language, this guarantee doesn't seem possible because cells can mutate objects and that seems impossible to prevent. Consider the following example:

Cell A:
```python
foo = {"key": "value"}
print(foo["key"])
```

Cell B:
```python
print(foo["key"])
```

Cell C:
```python
foo["key"] = "value 2"
print(foo["key"])
```

Now, if I run them in order, you get what you'd expect the first time around. The problem is if I go back to Cell B and run that cell again. I now get a different value than I did the first time (see screen shot).

**The bug part of this is issue that if I re-run Cell A and then Cell B, I'm still seeing "value 2" printed.**

But the larger point of this issue is that, unless I'm missing something, it can't be the case that marimo notebooks truly guarantee consistency and reproducibility in all cases? Even if we tracked some types of modification, you could never catch them all. Code can call into other libraries, code can dynamically eval string code, etc. I'd love to be wrong here because this would be awesome if we could reach the same level of guarantees that e.g. [Livebook](http

<img width="764" alt="Screenshot 2024-05-15 at 2 33 06 PM" src="https://github.com/marimo-team/marimo/assets/606233/16f04931-dd83-458d-915d-415ede81703b">
s://livebook.dev/) does, but I don't see how that's possible without using an immutable language.

### Environment

https://marimo.app (the online wasm version)

**Comment by akshayka** [comment]
> But the larger point of this issue is that, unless I'm missing something, it can't be the case that marimo notebooks truly guarantee consistency and reproducibility in all cases?

That's correct, and a good point. We try our best, without resorting to magic (ie, no runtime tracing, which is a losing battle), and are currently happy with being substantially more computationally reproducible than vanilla IPyKernel.

Our prescription to our users is: don't mutate variables across multiple cells. If you do, bets on "computational reproducibility" are off, unless you carefully manage the DAG yourself. You can find such language in our [docs](https://docs.marimo.io/guides/best_practices.html) and tutorials.

In this way we are similar to [`Pluto.jl`](https://github.com/fonsp/Pluto.jl?tab=readme-ov-file#dynamic-environment), which is our main inspiration. Pluto.jl takes reproducibility one step further than us by implementing its own package management system, which we have not yet done.

Happy to continue chatting. I will convert this to a Discussion because it's not bug, rather a design choice we're committed to.
